Document custom bundle process (#2615)
Shopify/ruby-lsp: jekyll/custom-bundle.markdown

---
layout: default
title: Custom Ruby LSP bundle
nav_order: 40
parent: Ruby LSP
---

# Custom Ruby LSP bundle

In language ecosystems other than Ruby, it is not super common to have to add editor tooling as part of your project

> dependencies. Usually, language servers are an executable that gets downloaded and then run independently from your

Reviewer: ```suggestion
dependencies. Usually, language servers are executables that gets downloaded and then run independently from your
```
or
```suggestion
dependencies. Usually, a language server is an executable that gets downloaded and then run independently from your
```

projects.

In the Ruby ecosystem, there are a few blockers to fully adopting this approach:

1. Not writing the language server in Ruby would make it challenging to integrate seamlessly with existing tools used by
the community that are already written in Ruby, like test frameworks, linters, formatters and so on
2. Discovering project dependencies automatically allows the language server to detect which files on disk must be read

> and indexed, so that we can extract declaration from gems without requiring any configuration from the user. This means

Reviewer: ```suggestion
and indexed, so that we can extract declarations from gems without requiring any configuration from the user. This means
```

> that we need to integrated directly with Bundler

Reviewer: ```suggestion
that we need to integrate directly with Bundler
```

> 3. Bundler only allows requiring gems that are setup as part of the `$LOAD_PATH`. If the Ruby LSP executable was running

Reviewer: ```suggestion
3. Bundler only allows requiring gems that are set up as part of the `$LOAD_PATH`. If the Ruby LSP executable was running
```

independently from a global installation, then the Ruby process would only be able to require the Ruby LSP's own
dependencies, but it would not be able to require any gems used by the project being worked on. Not being able to require
the project's dependencies limits integrations that the language server can automatically make with linters, formatters,
test frameworks and so on

> To overcome these limitations while at the same time not requiring users to add `ruby-lsp` as a dependency of their projects,

Reviewer: ```suggestion
To overcome these limitations, while at the same time not requiring users to add `ruby-lsp` as a dependency of their projects,
```

the Ruby LSP uses a custom bundle strategy. The flow of execution is as follows:

> 1. The executable is run `ruby-lsp` without `bundle exec` to indicate that the custom bundle must first be configured

Reviewer: ```suggestion
1. The executable is run as `ruby-lsp` without `bundle exec` to indicate that the custom bundle must first be configured
```
